fortanix-vme/fortanix-vme-runner/src/lib.rs

```diff
@@ -91,55 +91,43 @@ impl Listener {
-    fn new(vsock_stream: VsockStream<Std>, tcp_stream: TcpStream, remote_name: String) -> Self {
+    // Preliminary work for PLAT-367
+    fn new(_vsock_stream: &VsockStream<Std>, _tcp_stream: &TcpStream) -> Self {
         Connection {
-            tcp_stream,
-            vsock_stream,
-            remote_name,
         }
     }
 
-    fn close(&self) {
-        let _ = self.tcp_stream.shutdown(Shutdown::Both);
-        let _ = self.vsock_stream.shutdown(Shutdown::Both);
-    }
-
-    /// Exchanges messages between the remote server and enclave. Returns `true` when the
-    /// connection should remain active, false otherwise
-    fn proxy(&mut self) -> bool {
-        fn exchange<S: StreamConnection, D: StreamConnection>(src: &mut S, src_name: &str, dst: &mut D, dst_name: &str) -> bool {
-            // According to the `Read` threat documentation, reading 0 bytes
-            // indicates that the connection has been shutdown correctly. So we
-            // close the proxy service
-            // https://doc.rust-lang.org/std/io/trait.Read.html#tymethod.read
-            match Server::transfer_data(src, src_name, dst, dst_name) {
-                Ok(n) if n == 0 => false,
-                Ok(_)           => true,
-                Err(_)          => false,
-            }
-        }
-        let remote = self.tcp_stream.as_raw_fd();
-        let proxy = self.vsock_stream.as_raw_fd();
+    /// Exchanges messages between the remote server and enclave. Returns on error, or when one of
+    /// the connections terminated
+    fn proxy(tcp_stream: &mut TcpStream, vsock_stream: &mut VsockStream<Std>, remote_name: &str) -> Result<(), IoError> {
+        let remote = tcp_stream;
+        let proxy = vsock_stream;
 
-        let mut read_set = FdSet::new();
-        read_set.insert(remote);
-        read_set.insert(proxy);
-
-        if let Ok(_num) = select(None, Some(&mut read_set), None, None, None) {
-            if read_set.contains(remote) {
-                return exchange(&mut self.tcp_stream, &self.remote_name, &mut self.vsock_stream, "proxy");
-            }
-            if read_set.contains(proxy) {
-                return exchange(&mut self.vsock_stream, "proxy", &mut self.tcp_stream, &self.remote_name);
+        loop {
+            let mut read_set = FdSet::new();
+            read_set.insert(remote.as_raw_fd());
+            read_set.insert(proxy.as_raw_fd());
+
+            if let Ok(_num) = select(None, Some(&mut read_set), None, None, None) {
+                if read_set.contains(remote.as_raw_fd()) {
+                    // According to the `Read` trait documentation, reading 0 bytes
+                    // indicates that the connection has been shutdown correctly. So we
+                    // close the proxy service
+                    // https://doc.rust-lang.org/std/io/trait.Read.html#tymethod.read
+                    if Server::transfer_data(remote, remote_name, proxy, "proxy")? == 0 {
+                        return Ok(());
```

Comment:
Similar comment to the previous code: because we `return` here, I think we could in principle have starvation of the proxy --> remote side of the connection.